src/communication/expr.rs

```diff
@@ -0,0 +1,225 @@
+impl CommunicationReplacer for AssignExpr {
+    fn replace_communication_call(
+        &mut self,
+        modules: &Vec<ModuleComponent>,
+        module: &ModuleComponent,
+        class: Option<&ClassOrInterfaceComponent>,
+        method: &MethodComponent,
+    ) -> Option<Node> {
+        for expr in self.rhs.iter_mut() {
+            if let Some(Node::Expr(replacement)) =
+                expr.replace_communication_call(modules, module, class, method)
+            {
+                *expr = replacement;
+            }
+        }
+        None
+    }
+}
+
+impl CommunicationReplacer for CallExpr {
+    fn replace_communication_call(
+        &mut self,
+        modules: &Vec<ModuleComponent>,
+        module: &ModuleComponent,
+        callee_class: Option<&ClassOrInterfaceComponent>,
+        callee_method: &MethodComponent,
+    ) -> Option<Node> {
+        let mut endpoint_call = None;
+        let map_endpoint_call = |endpoint_call: Option<EndpointCallExpr>| {
+            endpoint_call.map_or_else(
+                || None,
+                |call| {
+                    let call: Expr = call.into();
+                    Some(call.into())
+                },
+            )
+        };
+
+        // Helper for placing an endpoint or RPC call into an Expr
+        let replace_node = |arg: &mut Expr| {
+            if let Some(Node::Expr(replacement)) =
+                arg.replace_communication_call(modules, module, callee_class, callee_method)
+            {
+                *arg = replacement;
+            }
+        };
+
+        // Client calls should be DotExpr's (e.g. compose_client->UploadUserId(...) in C++ or s.geoClient.Nearby(...) in Go)
+        let client_call = match &*self.name {
+            Expr::DotExpr(dot_expr) => Some(dot_expr),
+            _ => {
+                self.args.iter_mut().for_each(replace_node);
+                None
+            }
+        }?;
+
+        // Go a second level deep in the selector expression for Go's case
+        let client_call = match &*client_call.selected {
+            Expr::DotExpr(dot_expr) => Some(dot_expr),
+            _ => Some(client_call),
+        }?;
+
+        // Deduce information about the service being called by the client type or alternatively
+        // the client variable's naming convention
+        let client_ident = match_ident_or(&*client_call.expr)?.name.to_lowercase();
+        let method_ident = match_ident_or(&*client_call.selected)?;
+        let client_name = get_service_name(callee_class, &client_ident);
+        if client_name.is_none() {
+            self.args.iter_mut().for_each(replace_node);
+        }
+        let client_name = client_name?;
+
+        // Helper to see whether the endpoint has been found
+        let found_endpoint =
+            |class: &ClassOrInterfaceComponent, method: &MethodComponent, method_ident: &Ident| {
+                // Same method name and parameter count
+                let class_name_equiv = match callee_class {
```

Comment:
This variable is poorly named, will change it
```suggestion
let in_different_classes = match callee_class {
```

Reply:
Looks like a good change to me, especially since it's whether the names are not equal (as per the next line) and thus whether the classes are different not the same.